metal/resource_metal_virtual_circuit.go

```diff
@@ -0,0 +1,196 @@
+			},
+			"nni_vnid": {
+				Type:        schema.TypeInt,
+				Computed:    true,
+				Description: "Nni VLAN ID parameter, see https://metal.equinix.com/developers/docs/networking/fabric/",
+			},
+			"status": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Status of the virtual circuit resource",
+			},
+		},
+	}
+}
+
+func resourceMetalVirtualCircuitCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*packngo.Client)
+	vncr := packngo.VCCreateRequest{
+		VirtualNetworkID: d.Get("vlan_id").(string),
+		NniVLAN:          d.Get("nni_vlan").(int),
+		Name:             d.Get("name").(string),
+	}
+	connId := d.Get("connection_id").(string)
+	portId := d.Get("port_id").(string)
+	projectId := d.Get("project_id").(string)
+
+	conn, _, err := client.Connections.Get(connId, nil)
+	if err != nil {
+		return err
+	}
+	// TODO: export consts for Connection status from packngo
+	if conn.Status == packngo.VCStatusPending {
+		return fmt.Errorf("Connection request with name %s and ID %s wasn't approved yet", conn.Name, conn.ID)
+	}
+
+	vc, _, err := client.VirtualCircuits.Create(projectId, connId, portId, &vncr, nil)
+	if err != nil {
+		return err
+	}
+	createWaiter := getVCStateWaiter(
+		client,
+		vc.ID,
+		d.Timeout(schema.TimeoutCreate),
+		[]string{packngo.VCStatusActivating},
+		[]string{packngo.VCStatusActive},
+	)
+
+	_, err = createWaiter.WaitForState()
+	if err != nil {
+		return fmt.Errorf("Error waiting for virtual circuit %s to be created: %s", vc.ID, err.Error())
+	}
+
+	d.SetId(vc.ID)
+
+	return resourceMetalVirtualCircuitRead(d, meta)
+}
+
+func resourceMetalVirtualCircuitRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*packngo.Client)
+	vcId := d.Id()
+
+	vc, _, err := client.VirtualCircuits.Get(
+		vcId,
+		&packngo.GetOptions{Includes: []string{"project", "port", "virtual_network"}},
+	)
+	if err != nil {
+		return err
+	}
+
+	return setMap(d, map[string]interface{}{
+		//"connection_id": vc.Connection.ID,
+		"project_id": vc.Project.ID,
+		"port_id":    vc.Port.ID,
+		"vlan_id":    vc.VirtualNetwork.ID,
+		"status":     vc.Status,
+		"nni_vlan":   vc.NniVLAN,
+		"vnid":       vc.VNID,
+		"nni_vnid":   vc.NniVNID,
+		"name":       vc.Name,
+	})
+}
+
+func getVCStateWaiter(client *packngo.Client, id string, timeout time.Duration, pending, target []string) *resource.StateChangeConf {
+	return &resource.StateChangeConf{
+		Pending: pending,
+		Target:  target,
+		Refresh: func() (interface{}, string, error) {
+			vc, _, err := client.VirtualCircuits.Get(
+				id,
+				&packngo.GetOptions{Includes: []string{"project", "port", "virtual_network"}},
+			)
+			if err != nil {
+				return 0, "", err
+			}
+			return vc, vc.Status, nil
+		},
+		Timeout:                   timeout,
+		Delay:                     10 * time.Second,
+		MinTimeout:                5 * time.Second,
+		ContinuousTargetOccurence: 5,
```

Comment:
what does this do and is it enough?

Reply:
Ah, the continuoustagetoccurence is probably not necessary here. There's no jitter in state of EM resources.

Otherwise, the statechangeconf is a struct for polling for resource state. It's the old way to do it (before contexts).

I use it here to wait for the VC state on create (just like with devices). It's also used when deleting the VC because it's necessary to detach the VLAN (which you pass in creating request of the VC).